ejector/ejector.go

```diff
@@ -0,0 +1,205 @@
+	ejectionCriteriaTimeWindow time.Duration
+
+	// Used to convert validator IDs to validator addresses.
+	validatorIDToAddressCache eth.ValidatorIDToAddressConverter
+
+	// Used to look up the latest reference number.
+	referenceBlockProvider eth.ReferenceBlockProvider
+
+	// Used to fetch operator state.
+	operatorStateCache operatorstate.OperatorStateCache
+}
+
+// NewEjector creates a new Ejector.
+func NewEjector(
+	ctx context.Context,
+	logger logging.Logger,
+	config *EjectorConfig,
+	ejectionManager *ThreadedEjectionManager,
+	signingRateLookupV1 SigningRateLookup,
+	signingRateLookupV2 SigningRateLookup,
+	validatorIDToAddressCache eth.ValidatorIDToAddressConverter,
+) *Ejector {
+	e := &Ejector{
+		ctx:                        ctx,
+		logger:                     logger,
+		ejectionManager:            ejectionManager,
+		signingRateLookupV1:        signingRateLookupV1,
+		signingRateLookupV2:        signingRateLookupV2,
+		period:                     config.EjectionPeriod,
+		ejectionCriteriaTimeWindow: config.EjectionCriteriaTimeWindow,
+		validatorIDToAddressCache:  validatorIDToAddressCache,
+	}
+
+	go e.mainLoop()
+
+	return e
+}
+
+// The main loop periodically evaluates validators for ejection.
+func (e *Ejector) mainLoop() {
+	ticker := time.NewTicker(e.period)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-e.ctx.Done():
+			e.logger.Info("ejector shutting down")
+			return
+		case <-ticker.C:
+			e.evaluateValidators()
+		}
+	}
+}
+
+// evaluateValidators looks up signing rates and evaluates which validators should be ejected.
+func (e *Ejector) evaluateValidators() {
+
+	v1SigningRates, err := e.signingRateLookupV1.GetSigningRates(
+		e.ejectionCriteriaTimeWindow,
+		nil, // all quorums
+		ProtocolVersionV1,
+		true, // omit perfect signers if possible
+	)
+	if err != nil {
+		e.logger.Error("error looking up v1 signing rates", "error", err)
+		return
+	}
+
+	v2SigningRates, err := e.signingRateLookupV2.GetSigningRates(
+		e.ejectionCriteriaTimeWindow,
+		nil, // all quorums
+		ProtocolVersionV2,
+		true, // omit perfect signers if possible
+	)
+	if err != nil {
+		e.logger.Error("error looking up v2 signing rates", "error", err)
+		return
+	}
+
+	// Combine data from v1 and v2 lookups, since the validator is likely to cancel ejection if it is active in either.
+	signingRates, err := combineSigningRateSlices(v1SigningRates, v2SigningRates)
+	if err != nil {
+		e.logger.Error("error combining signing rates", "error", err)
+		return
+	}
+	sortByUnsignedBytesDescending(signingRates)
+
+	for _, signingRate := range signingRates {
+		e.evaluateValidator(signingRate)
+	}
+}
+
+// evaluateValidator evaluates a single validator's signing rate and decides whether to eject it.
+func (e *Ejector) evaluateValidator(signingRate *validator.ValidatorSigningRate) {
+	if !IsEjectable(signingRate) {
+		return
+	}
+
+	validatorID := core.OperatorID(signingRate.GetValidatorId()[:])
```

Comment:
looks like `signingRate.GetValidatorId()[:]` would panic if `GetValidatorId` returns `nil`.

Are we absolutely sure that that could never happen?

Reply:
I think we can trust it, since it comes from things we control. But added a check just in case.